rosidl_typesupport_introspection_cpp/resource/msg__type_support.cpp.template

```diff
@@ -39,6 +39,42 @@ namespace rosidl_typesupport_introspection_cpp
 {
 
 @[if spec.fields]@
+@[for field in spec.fields]@
+@[if not field.type.is_primitive_type() and field.type.is_array]@
+size_t size_function__@(spec.base_type.type)__@(field.name)(const void * untyped_member)
+{
+@[if field.type.array_size and not field.type.is_upper_bound]@
+  return @(field.type.array_size);
+@[else]@
+  const std::vector<@(field.type.pkg_name)::msg::@(field.type.type)> * member =
+    reinterpret_cast<const std::vector<@(field.type.pkg_name)::msg::@(field.type.type)> *>(untyped_member);
+  return member->size();
+@[end if]@
+}
+
+const void * get_function__@(spec.base_type.type)__@(field.name)(const void * untyped_member, size_t index)
+{
+@[if field.type.array_size and not field.type.is_upper_bound]@
+  const @(field.type.pkg_name)::msg::@(field.type.type) * member =
+    reinterpret_cast<const @(field.type.pkg_name)::msg::@(field.type.type) *>(untyped_member);
+@[else]@
+  const std::vector<@(field.type.pkg_name)::msg::@(field.type.type)> & member =
+    *reinterpret_cast<const std::vector<@(field.type.pkg_name)::msg::@(field.type.type)> *>(untyped_member);
+@[end if]@
+  return &member[index];
```

Comment:
The calling rmw implementation needs to make sure to only pass a valid index. It checks with `size_function` before or uses `resize_function` to increase the size as necessary.

Therefore I wouldn't add error handling in this function.